File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.py
class Solution:
    # First you need map
    # the outpur is list and you can not simply append to the list, you need to modify the child result
    # So let's use top down
    #  => define another function
    #  => define answer list
    #  => pass current string as arg
    # If digits == "", it's the time to add curr to answer
    def letterCombinations(self, digits: str) -> List[str]:
        self.answer = []
        self.phone_map = [None, None, ["a", "b", "c"], ["d", "e", "f"], ["g", "h", "i"], ["j", "k", "l"],
        ["m", "n", "o"], ["p", "q", "r", "s"], ["t", "u", "v"], ["w", "x", "y", "z"]]
        self.recursion(digits, "")
        return self.answer

    def recursion(self, digits, curr):
        if digits == "" and curr != "":
            self.answer.append(curr)
            return
        if digits == "":
            return 
        digit = int(digits[0])
        chars = self.phone_map[digit]
        for c in chars:
            self.recursion(digits[1:], curr + c)      

File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.java
class Solution {
    private char[][] phoneMap;
    private List<String>  answer;
    public List<String> letterCombinations(String digits) {
        phoneMap = initPhoneMap();
        answer = new ArrayList<>();
        backtrack(digits, new StringBuilder());
        return answer;
    }

    public char[][] initPhoneMap() {
        char[][] map = {
            {'0', '0', '0'},
            {'0', '0', '0'},
            {'a', 'b', 'c'},
            {'d', 'e', 'f'},
            {'g', 'h', 'i'},
            {'j', 'k', 'l'},
            {'m', 'n', 'o'},
            {'p', 'q', 'r', 's'},
            {'t', 'u', 'v'},
            {'w', 'x', 'y', 'z'}
        };
        return map;
    }
    public void backtrack(String digits, StringBuilder letters) {
        if (digits.length() == 0) {
            if (letters.length() > 0) {
                answer.add(letters.toString());
            }
            return;
        }
        int num = digits.charAt(0) - '0';
        char[] chars = phoneMap[num];
        for (char c : chars) {
            letters.append(c);
            String nextDigits = digits.substring(1);
            backtrack(nextDigits, letters);
            letters.deleteCharAt(letters.length() - 1);
        }
        
    }
    
}